File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
input=sys.stdin.readline

n,k=map(int,input().split())
bag=[[0,0]]
result=0
dp=[[0]*(k+1) for _ in range(n+1)]
for _ in range(n):# a=무게 b=가치
    a,b=map(int,input().split())
    bag.append([a,b])

for i in range(1,n+1):
    for j in range(1,k+1):
        if bag[i][0]<=j:
            dp[i][j]=max(dp[i-1][j],bag[i][1]+dp[i-1][j-bag[i][0]])
        else:
            dp[i][j]=dp[i-1][j]
    result=max(dp[i])
print(result)

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static int dp[][];
	static List<Bag> list;
	static int n,k,result;
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n=Integer.parseInt(st.nextToken());
		k=Integer.parseInt(st.nextToken());
		
		list = new ArrayList<>();
		list.add(new Bag(0,0));
		dp=new int[n+1][k+1];
		result=0;
		
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			list.add(new Bag(Integer.parseInt(st.nextToken()),Integer.parseInt(st.nextToken())));
		}
		for(int i=1;i<=n;i++) {
			Bag bag=list.get(i);
			for(int j=1;j<=k;j++) {
				dp[i][j]=dp[i-1][j];
				if(j>=bag.v) 
					dp[i][j]=Math.max(dp[i][j], dp[i-1][j-bag.v]+bag.c);
				result=Math.max(result, dp[i][j]);
			}
		}
		System.out.println(result);
		
	}
	static class Bag{
		int v,c;
		Bag(int v,int c){
			this.v=v;
			this.c=c;
		}
	}
}

